### PR TITLE
inference: fix fieldtype tfunc design

### LINKED ISSUE #31666: Internal error/MethodError in complex use case
With a certain application (actually in several cases) involving BAT.jl and a host of other packages, I get this error (reproducible):

```
Internal error: encountered unexpected error in runtime:
MethodError(f=typeof(Core.Compiler.fieldindex)(), args=(Random123.Philox4x{T, R} where R where T<:Union{UInt32, UInt64}, :ctr1, false), world=0x0000000000000eb9)
```

Apologies up front: I tried very hard, but wasn't able to create an MWE - as soon as I simplify things too much, the error goes away. On the other hand, I haven't been able to find a workaround that works for the actual (non-reduced) application either.

So here's a script (long, but as reduced as far as I could) "bat_julia_internal_error.jl", together with a Project and Manifest file that reproducibly causes the error: https://gist.github.com/oschulz/d2417a6259684861071ad098e3cafffa

Here's my system info:

```
julia> versioninfo()
Julia Version 1.1.0
Commit 80516ca202 (2019-01-21 21:24 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
```

OS: Ubuntu Linux 16.04, 64-bit

### PR DESCRIPTION
**Pull request by vtjnash**
fix #31666 and add some tests for nothrow computation

### CODE REVIEW

File: base/compiler/tfuncs.jl [hunk 1]
```diff
@@ -902,9 +927,9 @@ function fieldtype_tfunc(@nospecialize(s0), @nospecialize(name))
         ft = ftypes[fld]
     end
 
-    exact = (isa(s0, Const) || isType(s0)) && !has_free_typevars(s)
+    exactft = exact || !has_free_typevars(ft)
```
Comment (Keno) [inline base/compiler/tfuncs.jl]: Is `exact` enough here? What about types that are `==`, but not `===`?

Comment (Keno) [reply base/compiler/tfuncs.jl]: for `Const` I mean. I do agree that we can at least say `Type{ft}` with this condition.

File: base/compiler/tfuncs.jl [hunk 2]
```diff
@@ -815,17 +816,24 @@ function fieldtype_nothrow(@nospecialize(s0), @nospecialize(name))
...
+    s === Bottom && return false # always
+    return _fieldtype_nothrow(s, name)
 end
 
-function _fieldtype_nothrow(@nospecialize(u), name::Const)
+function _fieldtype_nothrow(@nospecialize(s), name::Const)
+    u = unwrap_unionall(s)
     if isa(u, Union)
-        return _fieldtype_nothrow(u.a, name) || _fieldtype_nothrow(u.b, name)
+        return _fieldtype_nothrow(u.a, name) && _fieldtype_nothrow(u.b, name)
```
Comment (JeffBezanson) [inline base/compiler/tfuncs.jl]: This is overly conservative --- for example `fieldtype(Union{Int,Tuple{Any}}, 1)` doesn't throw.

Comment (vtjnash) [reply base/compiler/tfuncs.jl]: Ah, OK, I need to start propagating the "exact" flag from instanceof_tfunc then

File: base/compiler/tfuncs.jl (hunk 1 above)
Comment (vtjnash) [reply base/compiler/tfuncs.jl]: Good point. I think `Const` is actually almost always the wrong answer here

**Review by JeffBezanson — APPROVED** [review]

**Comment by oschulz** [comment]
@ararslan ,   @JeffBezanson : Could this be backported to the upcoming v1.1.1 as well? Or is it too late for that?

**Comment by JeffBezanson** [comment]
Should be safe to backport, along with #31848.

**Comment by KristofferC** [comment]
1.1.1 is too far in the release process to add anything to.

**Comment by ararslan** [comment]
Yes, 1.1.1 is frozen; it is effectively finished and will be tagged within a few days.

**Comment by oschulz** [comment]
Sure, I understand, no worries.

**Comment by oschulz** [comment]
Would be the a backport candidate for the next Julia LTS release, when the time comes (1.0.x or 1.1.x)?

**Comment by ararslan** [comment]
There likely won't be any more 1.1.x releases, but the change will be in 1.2.0. (Indeed it should already be present in the release candidate.) As for 1.0.x, I see no reason not to backport it.

**Comment by oschulz** [comment]
Thanks for the info, @ararslan! An eventual backport to 1.0.x would be nice (and I get that 1.1 can't be maintained long term, with 1.2 and 1.3 in the pipeline).

Oh, and yes, we had no more trouble with this since v1.2-rc2.

**Pull request merged**